File: backend/app/models/bank_reconciliation.py
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


BankTransactionDirection = Literal["inflow", "outflow"]
BankMatchStatus = Literal["unmatched", "suggested", "matched", "ignored"]


class BankAccount(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    bank_account_id: str = Field(min_length=1, max_length=80)
    bank_name: str = Field(min_length=1, max_length=120)
    account_number_masked: str = Field(min_length=1, max_length=80)
    currency: str = Field(default="CNY", min_length=3, max_length=3)
    linked_account_code: str = Field(default="1002", min_length=1, max_length=32)
    enabled: bool = True


class BankStatementLineCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str = Field(default="default", min_length=1, max_length=64)
    bank_account_id: str = Field(min_length=1, max_length=80)
    transaction_date: str = Field(pattern=r"^\d{4}-\d{2}-\d{2}$")
    direction: BankTransactionDirection
    amount: Decimal = Field(gt=0, max_digits=16, decimal_places=2)
    currency: str = Field(default="CNY", min_length=3, max_length=3)
    counterparty_name: str = Field(default="", max_length=120)
    summary: str = Field(default="", max_length=200)
    bank_reference: str = Field(min_length=1, max_length=120)


class BankStatementLine(BankStatementLineCreate):
    statement_line_id: str
    imported_at: str
    match_status: BankMatchStatus = "unmatched"


class BankStatementImportResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_set_id: str
    imported_count: int
    duplicate_count: int
    lines: list[BankStatementLine]

File: frontend/src/types/bankReconciliation.ts
export type BankTransactionDirection = "inflow" | "outflow";
export type BankMatchStatus = "unmatched" | "suggested" | "matched" | "ignored";

export interface BankStatementLineCreate {
  account_set_id: string;
  bank_account_id: string;
  transaction_date: string;
  direction: BankTransactionDirection;
  amount: string | number;
  currency: string;
  counterparty_name: string;
  summary: string;
  bank_reference: string;
}

export interface BankStatementLine extends BankStatementLineCreate {
  statement_line_id: string;
  imported_at: string;
  match_status: BankMatchStatus;
}

export interface BankStatementImportRequest {
  account_set_id: string;
  lines: BankStatementLineCreate[];
}

export interface BankStatementImportResult {
  account_set_id: string;
  imported_count: number;
  duplicate_count: number;
  lines: BankStatementLine[];
}

export interface BankMatchCandidate {
  statement_line_id: string;
  journal_entry_id: string;
  journal_line_id: string;
  direction: BankTransactionDirection;
  score: number;
  reasons: string[];
  statement_date: string;
  journal_date: string;
  statement_amount: string | number;
  journal_amount: string | number;
  currency: string;
  counterparty_name: string;
  summary: string;
}

export interface BankMatchCandidateResponse {
  account_set_id: string;
  bank_account_id: string;
  period: string;
  minimum_score: number;
  candidates: BankMatchCandidate[];
}

export interface CashJournalLine {
  entry_id: string;
  line_id: string;
  entry_date: string;
  period: string;
  source_type: string;
  source_id: string;
  account_code: string;
  account_name: string;
  direction: "debit" | "credit";
  cash_direction: BankTransactionDirection;
  currency: string;
  original_amount: string | number;
  base_amount: string | number;
  summary: string;
}

export interface BankBalanceReconciliationStatement {
  account_set_id: string;
  bank_account_id: string;
  period: string;
  bank_balance: string | number;
  book_balance: string | number;
  bank_received_not_booked: string | number;
  bank_paid_not_booked: string | number;
  book_received_not_bank: string | number;
  book_paid_not_bank: string | number;
  adjusted_bank_balance: string | number;
  adjusted_book_balance: string | number;
  unmatched_statement_count: number;
  unmatched_journal_count: number;
  unmatched_statement_lines: BankStatementLine[];
  unmatched_journal_lines: CashJournalLine[];
}
